client/components/LearnerViewDirectiveSpec.js

```diff
@@ -0,0 +1,200 @@
+    // we need to flush both $interval and $timeout.
+    var flushIntervalAndTimeout = function(timeToFlush) {
+      $interval.flush(timeToFlush);
+      $timeout.flush(timeToFlush);
+    };
+
+    it("should only activate autosave once", function() {
+      for (var i = 0; i < $scope.questionIds.length; i++) {
+        if (i !== 0) {
+          // Default question index is 0, we don't want to
+          // navigate to question 0 again because navigateToQuestion
+          // will already have stored code for that question and fail the test.
+          $scope.navigateToQuestion(i);
+        }
+        expect(!$scope.autosaveOn).toBe(true);
+        var questionId = $scope.questionIds[i];
+        var question = QuestionDataService.getQuestion(questionId);
+        var starterCode = question.getStarterCode(LANGUAGE);
+        // There should be no code stored in localStorage
+        // before autosave is triggered.
+        expect(CodeStorageService.loadStoredCode(
+          questionId, LANGUAGE)).toEqual(null);
+        $scope.autosave();
+        checkAutosaveDetail(questionId, starterCode);
+        expect(CodeStorageService.loadStoredCode(
+          questionId, LANGUAGE)).toEqual(starterCode);
+      }
+    });
+
+    // Autosave is triggered. 5 seconds later, autosave text should be
+    // displayed for 1 second. Then, the autosave text should be gone but the
+    // autosave is still running. At 10 seconds, cachedCode will be compared
+    // to $scope.code, since they are the same, autosave will stop. Thus, at
+    // 10.1 seconds, autosave should be canceled and autosaveOn should be false.
+    var checkAutosaveDetail = function(questionId, starterCode) {
+      expect($scope.autosaveOn).toBe(true);
+      // Flush 4900 milliseconds -- time: 4.9s
+      expect($scope.autosaveTextIsDisplayed).toBe(false);
+      flushIntervalAndTimeout(AUTOSAVE_MILLISECONDS - DELTA_MILLISECONDS);
+      expect($scope.autosaveOn).toBe(true);
+      expect(CodeStorageService.loadStoredCode(
+        questionId, LANGUAGE)).toEqual(null);
+      // Flush 200 milliseconds -- time: 5.1s
+      flushIntervalAndTimeout(2 * DELTA_MILLISECONDS);
+      expect(CodeStorageService.loadStoredCode(
+        questionId, LANGUAGE)).toEqual(starterCode);
+      expect($scope.autosaveOn).toBe(true);
+      expect($scope.autosaveTextIsDisplayed).toBe(true);
+      // Flush 1000 milliseconds -- time: 6.1s
+      flushIntervalAndTimeout(SECONDS_TO_MILLISECONDS);
+      expect($scope.autosaveTextIsDisplayed).toBe(false);
+      // Flush 4000 milliseconds -- time: 10.1s
+      flushIntervalAndTimeout(4 * SECONDS_TO_MILLISECONDS);
+      expect($scope.autosaveTextIsDisplayed).toBe(false);
+      expect($scope.autosaveOn).toBe(false);
+    };
+
+    it("should store the latest code into localStorage", function() {
+      for (var i = 0; i < $scope.questionIds.length; i++) {
+        if (i !== 0) {
+          // Default question index is 0, we don't want to
```

Comment:
The default question index is 0, so we don't want to navigate to question 0 again as navigateToQuestion will already have stored code for the current question and will fail the test.